### PR TITLE
Implement prepare_environment/sources/version functions with some fixes

### PR DESCRIPTION
**Pull request by ajelinski**
The tool still can be run by calling `./main.py PACKAGE --dir DEST` but now it needs to use conda's python with conda-build installed (conda-build pip package is OK too but a few more packages need to be installed and it has some problems with python 3.8) as conda-build's rendering is used to get metadata. So the easiest way is to run it after `conda init`.

After running, the files `DEST/recipe/recipe_clobber.yaml`, `DEST/recipe/recipe_append.yaml`, `DEST/conda-env/condarc` are the most interesting ones to see the result plus checking the tags on `DEST/git_repos/*`.

To fully test environment preparing:
1. change some conda configurations (system or user ones; the environment ones don't matter as clean environment will be created and tested in tool),
2. call `conda config --show-sources`,
3. run the tool,
4. call `conda config --show-sources` again,
5. run the tool to restore configurations with `./main restore`,
6. call `conda config --show-sources` again.

Building the package can be tested through the commands printed by the tool, e.g.:
> To build the package in the prepared environment, run:
>   conda activate (...)/DEST/conda-env
>   conda build (...)/DEST/recipe

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+Options::
+
+    package_dir         Path to the directory containing ``meta.yaml`` of the package
+    --out=DIRECTORY     Use DIRECTORY to store generated files on which ``conda build`` should be run
+    -v, --verbose       Add more verbosity to output
+
+Preparing the working directory
+-------------------------------
+
+Due to problems with Conda's own repository management [TODO: what exactly?], ``conda-build-prepare`` clones the repository by itself and allows ``conda build`` to operate on the prefetched copy.
```
Comment (mithro) [inline README.rst]: "repository management" isn't a good name here. I would say "source code git"

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+TODO
+
+Open questions:
+
+#. How to force Conda to not use condarc?
+#. How to prepare the environment for subsequent use of ``conda build``?
+   Generate a script that will be run instead of ``conda build``?
+
+Cleaning up git tags
```
Comment (mithro) [inline README.rst]: "Cleaning up git tags" --> "Preparing git tags for better version descriptions?"

File: README.rst [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+While we use the information extracted from tags, the exact version is provided to ``recipe_append.yaml`` via the ``version`` key.
+
+The following version formats are supported:
+
+- ``vX.Y``,
+- ``vX.Y.Z``,
+- ``vX.Y.Z-rcQ``,
+- all of the above variations with punctuation replaced with dashes or underscores.
+
+If there are no tags denoting the version, ``v0.0`` is used instead.
```
Comment (mithro) [inline README.rst]: If no valid tags are found, a ``v0.0`` tag is created on the oldest commit in the repo.

File: conda_build_prepare/conda_cmds.py [hunk 4]
```diff
@@ -1,9 +1,93 @@
...
+import io
+import shutil
+import tempfile
+import sys
+
+from prepare import get_local_channels, get_package_condarc
+from git_helpers import git_checkout, git_clone, git_describe, \
+        git_rewrite_tags, git_add_tag, git_clone_relative_submodules
+
+def find(key, dictionary):
```
Comment (mithro) [inline conda_build_prepare/conda_cmds.py:16]: Most of these Python function should have doc strings.

File: conda_build_prepare/conda_cmds.py [hunk 5]
```diff
@@ -27,11 +111,254 @@ def create_env(package_dir):
...
+                    print('Git had problems adding extra.tags line:')
+                    print('\t"{}"'.format(line.replace('\n', '')))
+                    print()
+                    continue
+        print()
+
+def _call_conda_cmd_in_env(cmd_string, env_path):
+    assert os.path.isdir(env_path), env_path
+
+    return subprocess.check_output(f'conda run -p {env_path} {cmd_string}'
```
Comment (mithro) [inline conda_build_prepare/conda_cmds.py:164]: You probably don't want to use `subprocess.check_output` with arbitrary conda commands as there is complicated blocking issues around pipes.

You should instead create temporary files and use `stdout=<file>`. There is a bunch of complexity around `close_fds=True` and `pass_fds=()`...

File: README.rst (hunk 1 above)
Comment (ajelinski) [reply README.rst]: I only moved this line when I was updating `README.rst` with more detailed description of how it works now: https://github.com/antmicro/conda-build-prepare/pull/1/files#diff-88b99bb28683bd5b7e3a204826ead112R84

I'm not sure I understand what this first part should say. `Due to problems with Conda's own source code git` doesn't seem to be what you wanted.

Could you also give an example of those problems?

**Pull request merged**